### PR TITLE
add simple destructure plugin

### PR DESCRIPTION
**Pull request by erwinmombay**
There is currently a bug in closure compiler where when we do an object destructuring assignment the property access is not validated properly in closure conformance check phase. (reported in https://github.com/google/closure-compiler/issues/3500, internal b/144013285)

Ex:
```
requirement: {
  type: BANNED_PROPERTY
  error_message: 'Use ampdoc instead'
  value: 'Document.prototype.head'
}
```

This will be caught:
`const head = document.head;`

This will not be caught:
`const {head} = document;`

As a workaround until the issue is properly fixed in closure compiler we will attempt to rewrite all destructuring assignment to a normal property access **only in the check-types** phase. We also currently will **not rewrite** destructuring assignment in the parameter as this causes a type error since we will need to pull down the destructuring to the body which essentially changes the parameter signature (we will tackle this in a separate issue)

### CODE REVIEW

File: build-system/test-configs/conformance-config.textproto [hunk 1]
```diff
@@ -398,3 +398,20 @@ requirement: {
   value: 'DocumentType.prototype.after'
   value: 'CharacterData.prototype.after'
 }
+
+# Document: most of properties should be access via ampdoc for proper scoping.
```
Comment (erwinmombay) [inline build-system/test-configs/conformance-config.textproto]: @dvoytenko i temporarily added this to trigger the warnings, i'll remove this before merging

**Comment by erwinmombay** [comment]
@jridgewell i removed the handling of the parameter object destructuring case as it causes type errors when we rewrite. Figuring out a way to handle it better

File: build-system/babel-plugins/babel-plugin-transform-simple-object-destructure/test/fixtures/transform/variable-declarations/input.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+ * limitations under the License.
+ */
+
+const {head, body} = document;
+
+const {x} = get();
+
+function func() {
+  const {head, body} = document;
+  let {x, y, z} = get();
```
Comment (dvoytenko) [inline build-system/babel-plugins/babel-plugin-transform-simple-object-destructure/test/fixtures/transform/variable-declarations/input.js]: Should we also test things like `const {head: z} = document;`

Comment (erwinmombay) [reply build-system/babel-plugins/babel-plugin-transform-simple-object-destructure/test/fixtures/transform/variable-declarations/input.js]: good point

Comment (erwinmombay) [reply build-system/babel-plugins/babel-plugin-transform-simple-object-destructure/test/fixtures/transform/variable-declarations/input.js]: done.

**Comment by erwinmombay** [comment]
@jridgewell banned object destructuring assignment in params. added an eslint rule for it. i can add an auto fixer later if needed but since there was just a few locations i just manually fixed for now

File: build-system/eslint-rules/no-destructure-assignment-params.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+const selector =
+  'ObjectPattern[parent.type=FunctionDeclaration],' +
```
Comment (jridgewell) [inline build-system/eslint-rules/no-destructure-assignment-params.js]: Nit: `:function > ObjectPattern`: https://astexplorer.net/#/gist/451c5e119851b9a81fd0e7faf782c119/fc1387e2d03226e223c9d203e0c0292c24daf643

Comment (erwinmombay) [reply build-system/eslint-rules/no-destructure-assignment-params.js]: good catch

File: .eslintrc [hunk 4]
```diff
@@ -187,10 +188,12 @@
         "extensions/**/test/**/*.js",
         "extensions/**/test-e2e/*.js",
         "ads/**/test/**/*.js",
-        "testing/**/*.js"
+        "testing/**/*.js",
+        "build-system/**/*.js"
```
Comment (rsimha) [inline .eslintrc]: Adding `build-system/` here will cause all the other rules to be disabled as well. Instead, you can simply add `"local/no-destructure-assignment-params": 0,` to this list:

https://github.com/ampproject/amphtml/blob/23201dd241631754eec59a9d27d734d950bb88bb/build-system/.eslintrc#L21-L29

File: babel.config.js [hunk 5]
```diff
@@ -29,7 +29,7 @@ const minimist = require('minimist');
 const {isTravisBuild} = require('./build-system/common/travis');
 const argv = minimist(process.argv.slice(2));
 
-const isDist = argv._.includes('dist');
+const isDist = argv._.includes('dist') || argv._.includes('check-types');
```
Comment (rsimha) [inline babel.config.js]: Since this now includes `dist` and `check-types`, maybe rename `isDist` to `usesClosureCompiler`?

Comment (erwinmombay) [reply babel.config.js]: done.

File: .eslintrc (hunk 4 above)
Comment (erwinmombay) [reply .eslintrc]: done

**Pull request merged**